### PR TITLE
Feature/widget manager cleanup

### PR DESCRIPTION
**Pull request by dmolineus**
As discussed [here](https://github.com/contao-community-alliance/dc-general/pull/136#issuecomment-118076871) this PR extracts the widget building into an own subscriber to the BuildWidgetEvent.

It also introduce a new `PropertyValueEncoder` which extracts the encoding from the `ContaoWidgetManager`. This is done as the decoding is also required in the builder and this code is already [duplicated](https://github.com/contao-community-alliance/dc-general/blob/develop/src/ContaoCommunityAlliance/DcGeneral/Contao/Event/Subscriber.php#L148-L157).

There are still some open questions. I add them as comments to inline comments.

### CODE REVIEW

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/ContaoWidgetManager.php [hunk 1]
```diff
@@ -467,118 +212,23 @@ public function getWidget($property, PropertyValueBag $inputValues = null)
...
-        $propExtra['required'] = ($varValue == '') && !empty($propExtra['mandatory']);
-
-        if ($inputValues) {
-            $values = new PropertyValueBag($inputValues->getArrayCopy());
-            $model  = clone $this->model;
-            $model->setId($this->model->getId());
-            $this->environment->getController()->updateModelFromPropertyBag($model, $values);
-        } else {
-            $model = $this->model;
-        }
```
Comment (dmolineus) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/ContaoWidgetManager.php]: I think that the model cloning could be done before triggering the event. So not every own subscriber has to do it again (and then there is no need to pass the `PropertyValueBag` to the `BuildWidgetEvent` at all).

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php [hunk 2]
```diff
@@ -644,7 +644,8 @@ public function execute()
         $blnSubmitted            = ($inputProvider->getValue('FORM_SUBMIT') === $definition->getName());
         $blnIsAutoSubmit         = ($inputProvider->getValue('SUBMIT_TYPE') === 'auto');
 
-        $widgetManager = new ContaoWidgetManager($environment, $this->model);
+        $encoder       = new PropertyValueEncoder($environment);
```
Comment (dmolineus) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/EditMask.php]: I'm not very happy with the instantation of the encoder. I prefer defining it as service. But as the encoder only triggers the event there should not be any reasons to define another one. So I could live with that solution.

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/Subscriber/WidgetBuilder.php [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+        $event = new GetAttributesFromDcaEvent(
+            $arrConfig,
+            $property->getName(),
+            $varValue,
+            $propertyName,
+            $defName,
+            new DcCompat($environment, $model, $propertyName)
+        );
+
+        // FIXME: propagator
```
Comment (dmolineus) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/Subscriber/WidgetBuilder.php]: Afaik the propagator was removed from the code. What about the suffixed events, do they still exists or will they get removed as well?

Comment (discordier) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/Subscriber/WidgetBuilder.php]: Will get removed. A lot of them already have `FIXME` markings.

File: src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/Subscriber/WidgetBuilder.php [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+    /**
+     * Handle the build widget event.
+     * 
+     * @param BuildWidgetEvent $event The event.
+     * 
+     * @return void
+     */
+    public static function handleEvent(BuildWidgetEvent $event)
+    {
+        if ($event->getWidget()) {
```
Comment (discordier) [inline src/ContaoCommunityAlliance/DcGeneral/Contao/View/Contao2BackendView/Subscriber/WidgetBuilder.php]: IMHO we ALSO shouold put here:
```
if (TL_MODE !== 'BE') { 
    return;
}
```

**Comment by discordier** [comment]
Could you rebase your PR please? Will merge asap then.

**Comment by discordier** [comment]
Did it on my own, merged in b20f8def3d7871aef3cf67c298c6a344a76e9d60.

**Pull request closed without merging**